### DISCUSSION TITLE
Support offline installation of language server tools and extensions for remote SSH development

### DISCUSSION THREAD

**Discussion by mrzzmrzz** [zed-industries/zed#23614] [Feature Requests]
### Check for existing issues

- [X] Completed

## Current Situation
I'm working with a remote server that can only be accessed via SSH and has no internet connection due to security restrictions. This presents challenges when trying to set up Python development environment since tools like `node` and `pyright` cannot be directly installed on the remote server.

## Problem
Currently, it's impossible to use Python language server features (like code completion, type checking, etc.) on the remote server because:
1. The remote server has no internet access to download required dependencies
2. Tools like `node.js` and `pyright` cannot be installed directly
3. There's no built-in way to transfer and configure these tools from a local machine

## Feature Request
Would it be possible to implement a feature that allows:
1. Automatic downloading of required tools (node, pyright, etc.) on the local machine
2. Packaging these tools in a way that can be transferred to the remote server
3. Automatic configuration of these tools on the remote server through the SSH connection

This would enable developers working with air-gapped or restricted network environments to still utilize the full features of Python language server.

**Comment by notpeter** [comment]
Hi! While I understand the engineering scope of your request (it's pretty big) and the user-pain of operating in such environments, I feel like Zed may not be the correct tool to deliver that functionality.  Currently our SSH Remoting requires a single binary executable which communicates over SSH with Zed on your machine.  We expect in restrictive development environments, users will not even be allowed to write and execute arbitrary binaries the client and so zed-remote-server will instead be pre-installed on the host system.

Practically speaking though, you're asking for Zed to implement a platform independent method to bundle all the binaries and dependencies required for a development environment and deliver them to a remote machine where you expect them to be operable, without requiring root and to work no matter the platform/distribution/architecture.

While I appreciate the vote of confidence in Zed's ability to deliver useful things, I'm going to say for the moment that this is not something we plan to implement.  Thanks for reporting.

**Comment by mrzzmrzz** [comment]
Thank you very much for your response. Due to certain network conditions (China's Great Firewall or server firewall settings), I am currently unable to download Node.js, but I can access specific network resources. I am considering whether there is a manual way to inform us how to place these dependencies so that zed can properly call these programs? Perhaps in the absence of an internet connection, it would be possible to provide offline extension installation files like VSCode? Thank you very much for your hard work.

**Comment by notpeter** [comment]
Ah, I understand better now. In general things that are in your PATH will be used by the remote before it attempts to download them. For example, if `which pyright` returns a path of an installed binary we will use that. The node stuff (prettier, etc) is a little harder but not impossible.

Conceivably rather than just assuming Zed can "npm install" everything we could create a manifest that could assist you in creating a bundle of requirements which you could preinstall out-of-band so Zed has the tools it needs.

**Comment by jansol** [comment]
> create a manifest that could assist you in creating a bundle of requirements which you could preinstall out-of-band so Zed has the tools it needs

I wonder if this could also help with node-based LSP on NixOS? A lot of them are packaged as standalone tools and work fine with other editors, just not in Zed as Zed insists on running them "manually" with its own node runtime.

**Comment by leelele-cug** [comment]
Hi, there. Any progress on this issue?

**Comment by sunerpy** [comment]
Who can tell me how to install extensions with remote ssh server (no internet)...
